drivers/api/no_os_dma.c

```diff
@@ -76,8 +76,11 @@ static void default_sg_callback(void *context)
 		old_xfer->xfer_complete_cb(old_xfer, next_xfer,
 					   old_xfer->xfer_complete_ctx);
 
-	if (!list_size) {
+	if (!list_size && data->desc->irq_ctrl) {
 		no_os_irq_disable(data->desc->irq_ctrl, data->channel->irq_num);
+		no_os_irq_unregister_callback(data->desc->irq_ctrl,
```

Comment:
This isn't really required, since calling `no_os_irq_register_callback()` for the same event/handler (the DMA channel's specific interrupt number in this case) is supposed to just update the already registered IRQ handler, without allocating memory again. So I have not unregistered the callback in this case as an optimization (since the whole calloc/free sequence will take a bit longer than just disabling the interrupt).

I guess the best place to call `no_os_irq_unregister_callback()` would be in `no_os_dma_remove()`.

Reply:
Ok,  so stm32_irq needs to be fixed to not allocate then.

Reply:
looking at https://github.com/analogdevicesinc/no-OS/blob/df71b3f243e1ae83ba9702e2db548761801ec048/drivers/platform/stm32/stm32_irq.c#L287 it seems we would need to do a lot of book-keeping and tracking to do to  "just update the already registered IRQ handler" in stm32_irq_register_callback if we do it for all the peripheral types.
Looks like what would need to happen  is an unregister and register even if its behind the scene. 
do you see some other way?

do we really need to optimize here? this is a one time thing whenthe xfers are already completed right?

Reply:
Ok, you're right. I knew this was the case on Maxim and pico platforms and assumed STM32 to be the same. So we can keep you version then.

Reply:
looking at it more..and the examples in Maxim and pico.. perhaps we can factor out the allocation in stm32
ill give it a try.

Reply:
@CiprianRegus  the new push should address  your original comment.